### ISSUE TITLE
Lab 14: HOFs I

### ISSUE THREAD

**Issue by lmock** [cs10/bjc-r#436]
Please comment below to leave us your lab reviews! Remember, please follow this format:

```
Your Name:

Overview of Lab Comments: i.e. Was this lab good or bad? Too fast? Too Slow? Does it need something new?

For specific issues: (you should copy this section as many times as necessary)
Page Link:
Type of Issue: i.e. (code) bug, grammar, typo, lack of clarity, etc
Severity: i.e. "prevents students from continuing", "minor inconvenience", etc
Comments: Description of the problem, what can be done to fix it, etc.
```

**Comment by Frankgoji** [comment]
Franklin Lee

Overview of Lab Comments:
I like that this lab is very well-structured. It proceeds logically to show how HOFs are an important generalization and abstraction tool. Additionally, the structure of the starter page reflects the structure of the lab; separating each different component to another sprite works very well for this particular lab (though, there is the small issue of having the Pandigital sprite in front of the Factorion sprite, but that is trivial). The lab really emphasizes the fact that one overall block that can take functions as data can stand in for a whole family of blocks. I also think that it's valuable that for all the small helper blocks, the lab challenges students to use HOFs.

For specific issues:

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/functions-as-data.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=3
Type of Issue: clutter in starter file
Severity: mild
Comments: The problem isn't in the page, but in the starter file. The sprite HOFs has a huge number of blocks on the stage. Even though great care was taken in carefully arranging all the blocks in a logical, structured way, it is still too haphazard, especially to someone who isn't used to thinking of functions as data. The previous sprites used comments well, marking information valuable to the student. It might be worth adding a few more comments to this particular sprite.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/using-call-and-run.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=4
Type of Issue: Comprehension
Severity: trivial
Comments: This page did a great job in explaining the role of ringify, and when and why it should appear. However, I found the analogy that Dan Garcia employed to be the best way to remember when and where to use the grey rings: think of the grey rings as candy wrappers. You don't want to leave your candy exposed, so you wrap it. You also don't want to wrap it twice, since that's unnecessary.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/using-call-and-run.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=4
Type of Issue: lab structure
Severity: mild
Comments: It seems a bit confusing to ask students to make a new function, drawing a different type of line, and then telling them about how to make grey ring appear in your block. The way the paragraphed is structured, it would seem like the lab is asking students to make a new line function for the tree to take in, that itself takes in a block. There should at least be a new paragraph for the description of how to make your own HOFs blocks.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/generalizing-list-all.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=6
Type of Issue: code
Severity: mild-moderate
Comments: For the more generailized list-all block, the hint given says to use the call block. I've found, however, that it's simpler and easier to not use the call block at all. I know that it's important to familiarize students with the call block, but perhaps another example will work out. In this particular example, the call block hint might just confuse students. Also, the lab doesn't mention that these blocks are located in the HOFs sprite!

**Comment by tcabs17** [comment]
Tori Cabot

Overview of Lab Comments:
This is an engaging and informative lab. I enjoyed this when I was in the class. I like the "warm up problems" that challenge students to be self-aware and test themselves. I think we could perhaps incorporate one or two "review problems" at the start of most labs to refresh helpful skills. For the "functions as data" parts of the lab, perhaps someone should add an explicit statement somewhere making sure students are not just making an "if" case for the functions they think will be passed into the function as data. 

Specific Issues:
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/factorions.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=2 
Type of Issue: clarity
Severity: moderate
Comments: This slide is large even though it is only asking students to construct two blocks. Instead of using sentences, a more visual representation using arrows would make this slide less intimidating.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/functions-as-data.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=3
Type of Issue: clarity
Severity: moderate
Comments: This section is engaging new material and the lab instructions organize the concepts and blocks nicely. The SNAP! file, however, is cluttered. Perhaps some of the "call" examples should be put in another sprite to provide some clearer division of concepts.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/pandigital-numbers.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=5
Type of Issue: organization
Severity: minor
Comments: It might make more sense to have "pandigital" after "functions as data" and before "call and run".

**Comment by jeeves833** [comment]
Your Name: Lucas Karahadian

Overview of Lab Comments: I feel like the lab should make it more clear that scripts have been spread out between the different sprites in the starter file.  Other than that, this was a really well put together lab.  It does a good job of reinforcing the idea of HOFs as well as helping students to understand how functional programming looks.

For specific issues:

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/pandigital-numbers.html
Type of Issue: Misleading statement
Severity: potential confusion
Comments: The last sentence of the first paragraph refers to blocks “created on the last page.”  However, there were no blocks created on the last page.  This may be due to reordering of pages that the lab went through at some point.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/minimize-function.html
Type of Issue: Unclear directions
Severity: minor confusion
Comments: It should be made clear to students that they have to make the “f(n) = (1021 \* n) mod 101” block and that it was not given.

**Comment by chrisk867** [comment]
Name: Chris Kilian

Overview of lab: I really like this lab. The exercises are pretty interesting and kind of fun to work through, while at the same time not being too hard. The special names for cool numbers is also pretty cool to learn. I remember taking this lab when I took the class and the run and call blocks really confused me though.

Issues:
https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/using-call-and-run.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=4
Type of issue: clarity
Severity: medium
Comments: This page goes over how to use the call and run blocks very well, but it doesn't do a very good job explaining why you would use them. When I took the class, I knew the syntax of how to use these two block, but I didn't feel like I walked away from the lab knowing what type of problem I should be applying these blocks to.

**Comment by yifatamir** [comment]
Yifat Amir

Overview of Lab Comments: This lab is lots of fun and personally one of my favorites. It made me see computer science as much more powerful than I had previously viewed it. Also, this lab was really clear due to all of the examples and screenshots, so I didn't see too much that had to be changed.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/minimize-function.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=7 
Type of Issue: lack of clarity
Severity: mild (especially important for non-mathy students)
Comments: The part that says "finds the minimum of that function over those inputs" is not very clear... Students may get confused between inputs and outputs, and which function "that function" refers to is somewhat ambiguous. I would reword this sentence and then include a screenshot of an example of this being used. For example, put in the sqrt function and a list "1, 2, 3, 4" and show that the block should report 1.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/compose-over-list.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=8
Type of Issue: lack of clarity
Severity: minimal
Comments: The very last line is worded in a way that makes me think I'm supposed to edit the compose block, thought actually I am supposed to CLICK edit, so that I can see how it works. This should be reworded.

**Comment by finnhowell** [comment]
Name: Fionnuala Howell

Overview of Lab Comments: I think this lab is really helpful for preparing for the final and bringing together all of the different aspects of HOF's. The problems start easy and work up to challenging which is extremely helpful so that students don't get stuck on a hard problem and give up. The Functions As Data page was very clear and informative about using functions together. When I took CS10, I remember being confused by what ringify was and how to use it/how it made a difference in code, so I would emphasize that because it isn't a difficult concept but seems to come out of nowhere. The descriptions for ringify seem great in this lab!

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/hofs-practice-warmup.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=1
Type of Issue: lack of clarity
Severity: minor
Comments: I think it would be helpful to write with the starter file that the different scripts that they should be implementing are spread out in the starter file and set up for them. This lab is great for it's organization so it would be even better to let students know to use these different sprites!

**Comment by mfujimoto** [comment]
Your Name: Mason Fujimoto

Overview of Lab Comments: This was a good lab. I thought the beginning exercises were challenging but gave a lot of good practice with HOFs. My biggest concern would be that the minimize and compose pages are not satisfying in my opinion. The minimum just shows students that they can apply a minimum rule over a list while the compose page does not force students to apply their knowledge of compose.

For specific issues: 
Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/compose-over-list.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=8
Type of Issue: Hard to see if you are correct
Severity: minor/may allow students to be wrong and not know it
Comments: As the results from the compose block are always abstract (in terms of functions) students are not able to see if their compose block works as they expected. I suggest forcing the students to see if their composition is truly equivalent to what they expect by forcing them to use it in a block.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/minimize-function.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=7
Type of Issue: Flow
Severity: none
Comments: I thought this was just a strange/off topic thing to mention. Really it just shows students that they can use minimum as a function inside a HOF to find a minimum value. Worth knowing, but the exercise takes about 20 seconds to do and may be best included/worked into another slide or just left out.

**Comment by ghost** [comment]
Your Name: Anjli Shah

Overview of Lab Comments: i.e. Was this lab good or bad? Too fast? Too Slow? Does it need something new?

For specific issues: 
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/factorions.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=2
Type of Issue: Too wordy
Severity: Minor inconvenience
Comments: I noticed in the other labs that students often missed parts of the code because they skimmed over the requirements. If these were put in bullet points, or in a more readable form I think it would be clearer.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/functions-as-data.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=3
Type of Issue: Not enough emphasis
Severity: Limits understanding
Comments: I think the rings is a kind of tricky concept to understand, and a better/more in-depth introduction to them would help students in using them.

**Comment by amrutayel** [comment]
Your Name: Amruta Yelamanchili

Overview of Lab Comments: This was a good lab. I think it was well paced, and there was just enough information presented in it. They grey rings are kind of confusing throughout, they could use a little more explanation.

For specific issues: 

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/hofs-practice-warmup.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=1
Type of Issue: clarity
Severity: minor
Comments: It says the “list all numbers between 5 and 8” block should report (list 5 6 7 8), but this is a little bit confusing. Maybe that list should be replaced with a list in the format of the grey box. Same with the word -> list block. Importing the words, sentences library might be beneficial to students in creating this block.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/functions-as-data.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=3
Type of Issue: flow
Severity: minor
Comments: This page and the next seem a little out of place. Pandigital is right after, which relates more closely to factorions.

**Comment by erikdahlquist** [comment]
Erik Dahlquist

Overview: This lab has loads of good information, and can be enjoyable if you understand the HOFs we've already worked with. If you don't, even the first few slides will be a struggle.
I think pacing is fine, but I feel maybe we should present some easier "make your own HOFs" to start with, instead of using home-made HOFs in HOFs.
Confident students should find this lab to be a reasonable length, but I foresee this lab being a struggle for many other students. I am unsure how to avoid this.
Overall, fairly well done lab with the occasionally unclear prompt.

Issues:
Page Link: https://inst.eecs.berkeley.edu/~cs10/fa14/
Type of Issue: Page link?
Severity: Major
Comments: First off, am I blind, or is the lab not linked off the CS10 page yet? I got there off github...

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/factorions.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=2
Type of Issue: Sprite-switching clarification
Severity: Minor
Comments: You may want to remind them they should keep switching Sprites depending on what slide they're on. Some of them seem to struggle with identifying they can do that.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/factorions.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=2
Type of Issue: Block-naming confusion
Severity: Moderate
Comments: Why have them create a "WORD --> list" block with a text input slot if they're going to be using it for numbers? Do you want them to make the logical connection that the block could be used for both? Because that's fine, it just seems potentially misleading to have them create it and name it with the impression it should be used for strings.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/using-call-and-run.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=4
Type of Issue: Prompt inconsistency
Severity: Moderate
Comments: The last paragraph is confusing:
"Try it out with the functions provided in the starter file, and then try making your own function that draws a different type of line. When making your own HOF blocks, you can get the gray-border input by clicking the input name in the block editor, then clicking the dropdown arrow. You'll be able to select "command" or "reporter" as the input type."
The first sentence seems potentially disconnected from the rest. This could be interpreted as they just need to ringify a new line-drawing function (to replace 'zig-zag' or 'simple line') inside the pre-made fractal function, so the fractal drawn is using different lines  (that's how I interpreted it...). But this doesn't necessitate making their own HOF at all. Figuring out whats actually being asked here is a bit confusing.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/pandigital-numbers.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=5
Type of Issue: Misleading prompt
Severity: Major
Comments: It's impossible to make the "display nicely" block without recursion or iteration, right? I think that should be clarified so they're not struggling with HOFing it.
Also, why are they even doing this block here? It's so random.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/generalizing-list-all.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=6
Type of Issue: Awkward wording
Severity: Minor
Comments: The first sentence is awkward considering the "list all pandigital numbers" block was written only last slide.

Page Link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/generalizing-list-all.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=6
Type of Issue: Confusing prompt
Severity: Moderate
Comments: Are they expected to create the "is _ even" block and use it in their new "list all numbers with property" block? Because if not, why did we give them a blank "is _ even" block? And if so, "you don't actually have to solve this one" should be rephrased to be more clear on what it's talking about, and they should be reminded they have a blank "is _ even" they can work off.

**Comment by judyworkman** [comment]
Judy Workman
Lab #14 Review (working with data)
Overview of Lab Comments: 
This lab is very interesting and relevant to modern times, but it appears as somewhat random and not as cohesive as most of the other labs are.  Perhaps, the connection between working with data and the rest of the course could be clarified in more depth so that students don’t feel like this lab is random or pointless.
For specific issues: 
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/data/spam-ham/1-introduction.html?topic=berkeley_bjc/areas/data.topic&step=0&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html
Type of Issue: lack of clarity
Severity: minor inconvenience
Comments: confusing definition of HAM
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/data/gps-digital-forensics/conclusions.html?topic=berkeley_bjc%2Fareas%2Fdata.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=10
Type of Issue: typo
Severity: minor inconvenience
Comments: “and returns how many points lie within in.”

**Comment by judyworkman** [comment]
Judy Workman
Lab #14 Review-working with HOFs
Overview of Lab Comments: 
This lab is very interesting and applicable to students’ understanding of HOFs.  The review of HOFs in the beginning of the lab is essential to help remind students of the basics before diving into more challenging problems and ideas.
For specific issues: 
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/pandigital-numbers.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=5
Type of Issue: lack of clarity
Severity: minor inconvenience
Comments: include more explaining why 3 is not pandigital
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/compose-over-list.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=8
Type of Issue: lack of clarity
Severity: minor inconvenience
Comments: somewhat confusing having the instructions that are broken up by code in between nearly every line

**Comment by carynbear** [comment]
Caryn Tran

Overview of Lab Comments: The lab is good overall. Still a bit wordy for my tastes. An alternate way to explain ringifying/call or run is that the ring pauses the block and then the call or run unpauses the block. 

For specific issues: (you should copy this section as many times as necessary)
Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/using-call-and-run.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=4
Type of Issue: wordiness
Severity: minor inconvenience
Comments: "The only difference between call and run is that call is for reporter blocks, and run is for command blocks." can be stated simply as: "Call is used with reporter blocks while run is used with command blocks."

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/minimize-function.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=7
Type of Issue: Lack of clarity
Severity: mild inconvenience
Comments: I think this assignment can be explained better. "The minimize block takes in a function (block with one input) and input values (list). It will report the minimum of all the answers generated by plugging each input value (list item) into the function." And then a visual step by step example instead of the one line example currently there.

**Comment by melvinhernandez** [comment]
Melvin Hernandez
Lab 14 Practice with HOFs and Functions as Data

Overview of Lab: I thought it was a great lab! It teaches a lot of problem solving and it allows students to think. I liked how each problem set would build from facts of the previous problem/sprite. I thought the timing was fine; it may take a little more than a two hours, but overall, I think the lab is the perfect length. One minimal suggestion would probably be to just add some sort of conclusion at the end of the lab to wrap-up the ideas seen.

Page link: http://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/pandigital-numbers.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=5
Type of issue: Confusing/Misleading Problem
Severity: Medium/Major
Comments: When creating the "display nicely" block, I couldn't think of a way to not add a comma to the last item ONLY using high order functions. I think it is a problem that would be easier (or only) done iteratively or even recursively.

Page link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/compose-over-list.html?topic=berkeley_bjc/hofs/hofs-practice.topic&step=8&novideo=true&noreading=true&noassignment=true&course=cs10_sp14.html
Type of issue: Bad Block
Severity: Major Inconvenience
Comments: Even though the compose section is supposed to be extra, it may bother someone who is trying to do it. I was not able to get the given "compose" block to work. The block only outputs its reporter in picture format. I think the idea was there but it may be a Snap bug. Because of this, students will not be able to see if their "compose from list" block works.

**Comment by jeffz6** [comment]
Your Name: Jeffrey Zhang

Overview of Lab Comments: This is a really fun lab.  Splitting all the parts into sprites and labelling everything that the students should or should not edit made the instructions really clear.  If I had to change one thing about it, it would be the functions chosen (specifically factorian).  There are such few factorians that the students have to run it on about 150 numbers just to get a nontrivial result.  It would save a lot of time to use something else, like magic numbers.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/generalizing-list-all.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassingment=true&course=cs10_fa14.html&step=6
Type of Issue: clarity
Severity: minor
Comments: Say that this is in the HOFs sprite

**Comment by strollinghome** [comment]
Carlos Flores

Overview:
This lab has a great structure. I like the way we start by solving a simple problem with HOF's and built it up to get a cool functionality that uses our initial blocks. The consistency of the structure makes for an easy going lab. Judging by what I saw in lab this week most students understood what was going on and were able to get through most of it in time. The last part is a bit confusing and I found some students to not use their compose block to implement the compose (list) block. I'm not sure if this was a general confusion or just a coincidence. Other than that this lab seems to be pretty solid.

**Comment by maknegana** [comment]
Your Name: Michelle Han

Overview of Lab Comments:I thought that was a nice, simplistic introduction to the better understanding of applying HOFs. I liked that the first page told the students what blocks they would be using. It made it simplistic for people to break down how to create the code for the other blocks. Usually most students understand the concepts but just don't know how to get started, and I felt that this lab took this into consideration.

For specific issues & how to fix: I personally think that the linked page below and the page after should come after the min value of a list of numbers excercise. That would make the lab flow better.
Page Link:https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/functions-as-data.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassingment=true&course=cs10_fa14.html&step=3
Type of Issue: slight confusion
Severity: minor inconvenience

**Comment by LithiumH** [comment]
Your Name: Haojun Li

Overview of Lab Comments: Overall it is a pretty hard lab because most students have difficulties to understand that functions can be passed in as data. But this lab handles them well.

For specific issues: (you should copy this section as many times as necessary)
Page Link: http://beautyjoy.github.io/bjc-r/cur/programming/HOFs/functions-as-data/factorions.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=2
Type of Issue: Hard
Severity: Medium
Comments: This question is pretty hard to write factorial using hof. Maybe give some more hint like use combine block or use the list all numbers from block. Since this is just a warm up not the purpose of the lab.

Page Link: http://beautyjoy.github.io/bjc-r/cur/programming/HOFs/functions-as-data/functions-as-data.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=3
Type of Issue: Unclear
Severity: Medium
Comments: Suggest to put the "not be evaluated" be bolded for clarification.

Page Link: http://beautyjoy.github.io/bjc-r/cur/programming/HOFs/functions-as-data/pandigital-numbers.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=5
Type of Issue: Unclear
Severity: High
Comments: I do not know why the blocks created on the last page is useful in anyway. Maybe give more hints?

**Comment by rxia752** [comment]
Rowena Xia
Overview of Lab Comments: This lab is incredibly important, and does a good job of providing practice with HOFs. The structure, the problems, everything overall is very good. However, it's also very long, and I might consider it too long. I could definitely see some students struggling with the length of this lab.  It might be good to trim it a bit and get to the essentials? It is easy for students to get frustrated in this lab if they get stuck on something, which would be detrimental to the learning from this lab.  I think it would be a better learning experience to have fewer problems, but to help the students understand them as well as possible. 

There were not any particular specific issues that I noticed.

**Comment by Hungukaroo** [comment]
Elizabeth Steger
Overview: A very nice lab that is fundamental to understanding HOFs. I think the warm-ups make the lab a little less daunting :). I very much appreciate the organization of the sprites and think this is a very well composed lab. 

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/factorions.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=2

Type of Issue: Ordering
Severity: Mild
Comments: Maybe students should read the whole page before attacking a problem, but sometimes you really just want to go for it, and with no mention of factorial before the factorion? block, I think some students may try to jump straight into solving the predicate before implementing the reporter. I realize this isn't a big deal and I like the flow of the problem, but even adding a picture of the factorial block before the factorion explanation (or after) with the "To do this, you'll need the factorial block. This time, write it using HOFs, rather than recursion." might make this more clear.

Page Link: https://inst.eecs.berkeley.edu/~cs10/labs/cur/programming/HOFs/functions-as-data/pandigital-numbers.html?topic=berkeley_bjc%2Fhofs%2Fhofs-practice.topic&novideo=true&noreading=true&noassignment=true&course=cs10_fa14.html&step=5

Type of Issue: Incorrect
Severity: Mild to pretty confusing
Comments: There wasn't a block we created on the last page that helps us and this could really confuse some students struggling to solve the problem.

**Comment by MicBrain** [comment]
Rafayel Mkrtchyan

Overview of lab content: This is one of the most important labs for this semester. Understanding the content of this lab is also very helpful for CS10 final exam's preparation. Lab is well structured and is starts from easier problems. It is a very good idea to explain big ideas by demonstrating to use them to solve easy problems(easy problems are always a good practice). The lab is hard but understanding the concepts is vital for all students.
